File: 다이내믹 프로그래밍/1149.RGB거리.js
//[실버1] 1149.RGB거리 https://www.acmicpc.net/problem/1149


import { createRequire } from "module";
const require = createRequire(import.meta.url);

const fs = require("fs")
// let input = fs.readFileSync('input.txt').toString().split('\n');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');
const N = +input[0];

for (let i = 1; i <= N; ++i){
    input[i] = input[i].split(' ').map((elem) => +elem)
}

input.splice(0,1);

function solution(n, data){

    let dp = [];
    dp.push(data[0]);
    for (let i = 1; i < n; i++){
        let r = Math.min(dp[i-1][1], dp[i-1][2]) + data[i][0];
        let g = Math.min(dp[i-1][0], dp[i-1][2]) + data[i][1];
        let b = Math.min(dp[i-1][0], dp[i-1][1]) + data[i][2];
        dp.push([r,g,b]);
    }

    console.log(Math.min(...dp.at(-1)));
}

solution(N, input);

File: 다이내믹 프로그래밍/1149.RGB거리.cc
//[실버1] 1149.RGB거리 https://www.acmicpc.net/problem/1149

#include <iostream>
#include <string>
#include <climits>

using namespace std;

int n;
int rgb[1001][3];
int dp[1001][3];

void getInput(){

    cin>>n;
    int r, g, b;
    for(int i=0;i<n;++i){
        cin>>rgb[i][0]>>rgb[i][1]>>rgb[i][2];
    }
    
}

int main(void){

    getInput();

    for(int i=0;i<3;++i){
        dp[0][i] = rgb[0][i];
    }

    for(int i=1;i<n;++i){
        dp[i][0] = min(dp[i-1][1], dp[i-1][2]) + rgb[i][0];
        dp[i][1] = min(dp[i-1][2], dp[i-1][0]) + rgb[i][1];
        dp[i][2] = min(dp[i-1][0], dp[i-1][1]) + rgb[i][2];
    }

    int answer = INT_MAX;
    for(int i=0;i<3;++i){
        answer = min(answer, dp[n-1][i]);
    }
    cout<<answer;
}